### ISSUE TITLE
buggy behavior with events

### ISSUE THREAD

**Issue by alidcastano** [bustlelabs/ember-mobiledoc-editor#112]
Right now I have my mobiledoc saving changes when the user focuses out of the element.

I'm trying to add a separate input field for the title. But it seems that having another text area along with the editor causes buggy behavior. When I try to click on the title area, the mouse moves to the mobiledoc.

I've tried placing the input field outside the element but that didn't change anything. And neither do the preventDefault=false or bubbles=false options for events.

Questions:

- Is there a way to add separate action handlers (besides on-change) to the mobiledoc using this addon?
- Is there a better way to save the mobiledoc? (Using the on-change handler caused the mouse to move to the beginning on every change, which is why I resorted to 'focusOut'. But now it seems that I can't add another input field.)

**Comment by alidcastano** [comment]
UPDATE: the issue above was my fault, I was saving the editor on 'focusOut', which was preventing other actions inside the same component from taking place.

I figured out how to pass the updated mobiledoc as argument to the debounce function, as shown below, so that I could successfully implement autosave.

```

export default Ember.Component.extend({
  triggerSave: function(body) {
    this.sendAction('save', body); // sets and saves mobiledoc
  },

  actions: {
    bodyWasUpdated(changed) { // trigged on change
      Ember.run.debounce(this, this.get('triggerSave'), changed, 1000);
    }
  }
});

```

**But main problem still persist: whenever the editor is saved, the cursor position moves to the beginning of the document.**  Is there a way to prevent this using ember-mobiledoc?

**Comment by bantic** [comment]
@alidcastano Is there something happening in the `save` action that causes the editor's `mobiledoc` property to be updated? The behavior you're describing sounds like the editor is being rerendered and autofocused (which would put the cursor at the start of the editor). Bustle has an internal tool that does exacty what you describe (separate input for post title, alongside mobiledoc editor for post body) without issue. I will see if I can create a small example for you. In the meantime, can you share more of your code to help shed light on this issue?

**Comment by alidcastano** [comment]
I just created a [fiddle](https://ember-twiddle.com/4df12f81998e7ae6c2360400885b32fa?openFiles=templates.editor.hbs%2Ctemplates.components.note-body.hbs) with basically the exact same code. Not sure how to get it to actually save the record though.

The problem is that when the mobiledoc record succesfully saves, the cursor moves to the start position and messes up the typing.  Debouncing the calls only delays this behavior, it doesn't prevent it.

**Comment by bantic** [comment]
@alidcastano Thanks for putting together that reproduction.
I think the issue is that the note's `body` is getting `set`, and since the mobiledoc renderer's `mobiledoc` property is bound to the note's `body`, it sees a property change notification and re-renders.
I am going to try to make a smaller more focused reproduction. If this is what is happening we can fix it by changing the mobiledoc renderer to be smarter about not rerendering when the mobiledoc is `set` somewhere but its content doesn't actually change.

**Comment by alidcastano** [comment]
To be precise, I can set the note `body` and not save it, and everything is fine. The issue arises when the note `body` is saved.

**Comment by bantic** [comment]
@alidcastano Thanks for explaining.  I think I see what is happening now.

If you have the editor component in a template like so:
`{{mobiledoc-editor mobiledoc=note.body}}`

Any time `note.body` changes to a new object, the editor will rerender which causes the cursor to jump back to the beginning.

When you save the note, your data layer (ember-data, probably?) is probably returning a new `body` object in its response, and so the `note.body` object reference changes (even if the actual content of the mobiledoc hasn't changed, the object reference is different), and the editor will rerender.

I made an example app that showcases what I think is happening here:
https://github.com/bantic/try-mobiledoc-editor/tree/eme-issue-112
Diff here: https://github.com/bantic/try-mobiledoc-editor/compare/eme-issue-112

The index route has the behavior you describe because in the `on-change` hook it creates a new object `Ember.copy(mobiledoc, true)` and sets the mobiledoc property to that new value. This causes the editor to rerender.

What you probably want to do is not have that binding between the note's `body` and the `mobiledoc` property on the editor. The easiest way to do this is to use `unbound`. You can see on the example app above how that looks in the `unbound` route/controller/template. For your app it'd be something like:
`{{mobiledoc-editor mobiledoc=(unbound note.body)}}`.
Then you can still use the `on-change` to save updates that the user makes, but because you aren't also re-setting the editor's `mobiledoc` property every time, it won't rerender.

**Comment by alidcastano** [comment]
Wow, thank you. Adding `unbound` got it to work! 

I figured the problem was that the mobiledoc was being rerendered. I tried debouncing calls, saving on mouseLeave, buffering, etc. to prevent the action from bubbling up. Didn't know about the `unbound` helper.

Is this the how most people save the mobiledoc (or was I not meant to set `note.body` directly as the mobiledoc in the first place)? If so, do you think it would be helpful to add to the README? I saw the debounce suggestion there, which is why I had previously tried that as my first solution.

**Comment by rlivsey** [comment]
Should/could setting a new document & re-rendering maintain the cursor position?

We have this bit of code in our app to maintain cursor position, we call `stashCursor` in `willRender` and `reapplyCursor` in `setupEditor`:

```javascript
function stashCursor(editor) {
  if (!editor || !editor.cursor.hasCursor()) {
    return;
  }

  const { offset, section } = editor.cursor.offsets.head;
  const index = editor.post.sections.toArray().indexOf(section);
  return { index, offset };
}

function reapplyCursor(editor, selection) {
  if (!editor || !selection) {
    return;
  }

  const { index, offset } = selection;

  const section = editor.post.sections.objectAt(index);
  if (section) {
    const position = new Position(section, offset);
    const range    = new Range(position, position);
    editor.cursor.selectRange(range);
  } else {
    editor.element.focus();
  }
}
```

I wonder if something similar (and more robust) could be built in?

**Comment by bantic** [comment]
> Is this the how most people save the mobiledoc (or was I not meant to set note.body directly as the mobiledoc in the first place)? If so, do you think it would be helpful to add to the README? I saw the debounce suggestion there, which is why I had previously tried that as my first solution.

@alidcastano Yes, I think that's an excellent idea, and thanks for your patience as we figured that out.
Generally we've envisioned (and used) ember-mobiledoc-editor as a sort of "document generator" insofar as we use the `on-change` hook to shoot out updated/changed documents that then move on into the rest of the machinery of the app (i.e., get persisted to the backend, at some point). In ember looping in that mobiledoc property to the ember-data model's property where it is persisted is a natural pattern. The docs should absolutely point that out. If possible we should also make that pattern easier to use with this addon, too. Perhaps the cursor stabilization is a good first step.

@rlivsey Yeah — a heuristic to stabilize the cursor over re-renders seems like a good idea. I would happily take a PR for that. :) I'll create a tracking issue for it now and work on getting it in soon, regardless.
